JavaVersionDetector class fails with NumberFormatException on Temurin JDK 23 EA builds

The JavaVersionDetector class fails on specific version strings that `java.version` property can hold. In this case it's `23-beta`, which results in:

```
Caused by: java.lang.NumberFormatException: For input string: "23-beta"
	at java.base@23-beta/java.lang.Integer.parseInt(Integer.java:588)
	at java.base@23-beta/java.lang.Integer.parseInt(Integer.java:685)
	at org.awaitility.core.JavaVersionDetector.getJavaMajorVersion(JavaVersionDetector.java:21)
	at org.awaitility.core.LambdaErrorMessageGenerator.getLambdaDetectionClassName(LambdaErrorMessageGenerator.java:86)
	at org.awaitility.core.LambdaErrorMessageGenerator.isLambdaClass(LambdaErrorMessageGenerator.java:30)
	at org.awaitility.core.AssertionCondition.generateDescriptionPrefix(AssertionCondition.java:87)
	at org.awaitility.core.AssertionCondition.getMatchMessage(AssertionCondition.java:73)
	at org.awaitility.core.AssertionCondition.lambda$new$0(AssertionCondition.java:54)
	at org.awaitility.core.ConditionAwaiter$ConditionPoller.call(ConditionAwaiter.java:248)
	at org.awaitility.core.ConditionAwaiter$ConditionPoller.call(ConditionAwaiter.java:235)
```

Consider using the `Runtime.Version` API which is available for JDK 11+

```
jshell> var v = Runtime.getRuntime().version();
v ==> 23-beta+14-ea

jshell> v.feature()
$3 ==> 23
```

**Steps to reproduce:**
1. Download a JDK 23 EA build from Adoptium. Example here: https://adoptium.net/temurin/nightly/?version=23
2. Run `JavaVersionDetector.getJavaMajorVersion()` on it.
3. Observe `NumberFormatException`

**Aditional Info:**
The existing code seems to not account for the `-` in `23-beta`. See:
https://github.com/awaitility/awaitility/blob/fbe16add874b4260dd240108304d5c0be84eabc8/awaitility/src/main/java/org/awaitility/core/JavaVersionDetector.java#L6-L22

--- a/awaitility/src/main/java/org/awaitility/core/JavaVersionDetector.java
+++ b/awaitility/src/main/java/org/awaitility/core/JavaVersionDetector.java
@@ -14,6 +14,8 @@ public final class JavaVersionDetector {
             normalizedJavaVersion = "8";
         } else if (javaVersion.startsWith("1.")) {
             normalizedJavaVersion = javaVersion.substring(2, 3);
+        } else if (javaVersion.endsWith("-ea")) {
+            normalizedJavaVersion = javaVersion.substring(0, javaVersion.length() - 3);
         } else {
             int dot = javaVersion.indexOf(".");
             if (dot != -1) {

--- a/awaitility/src/main/java/org/awaitility/core/JavaVersionDetector.java
+++ b/awaitility/src/main/java/org/awaitility/core/JavaVersionDetector.java
@@ -4,18 +4,22 @@ public final class JavaVersionDetector {
     private static final String JAVA_VERSION = System.getProperty("java.version");
 
     public static int getJavaMajorVersion() {
+        return getJavaMajorVersion(JAVA_VERSION);
+    }
+
+    static int getJavaMajorVersion(String javaVersion) {
         final String normalizedJavaVersion;
-        if (JAVA_VERSION == null || JAVA_VERSION.isEmpty()) {
+        if (javaVersion == null || javaVersion.isEmpty()) {
             // Fallback to java 8
             normalizedJavaVersion = "8";
-        } else if (JAVA_VERSION.startsWith("1.")) {
-            normalizedJavaVersion = JAVA_VERSION.substring(2, 3);
+        } else if (javaVersion.startsWith("1.")) {
+            normalizedJavaVersion = javaVersion.substring(2, 3);
         } else {
-            int dot = JAVA_VERSION.indexOf(".");
+            int dot = javaVersion.indexOf(".");
             if (dot != -1) {
-                normalizedJavaVersion = JAVA_VERSION.substring(0, dot);
+                normalizedJavaVersion = javaVersion.substring(0, dot);
             } else {
-                normalizedJavaVersion = JAVA_VERSION;
+                normalizedJavaVersion = javaVersion;
             }
         }
         return Integer.parseInt(normalizedJavaVersion);

--- a/awaitility/src/main/java/org/awaitility/core/JavaVersionDetector.java
+++ b/awaitility/src/main/java/org/awaitility/core/JavaVersionDetector.java
@@ -14,15 +14,12 @@ public final class JavaVersionDetector {
             normalizedJavaVersion = "8";
         } else if (javaVersion.startsWith("1.")) {
             normalizedJavaVersion = javaVersion.substring(2, 3);
-        } else if (javaVersion.endsWith("-ea")) {
-            normalizedJavaVersion = javaVersion.substring(0, javaVersion.length() - 3);
+        } else if (javaVersion.indexOf('.') != -1) {
+            normalizedJavaVersion = javaVersion.substring(0, javaVersion.indexOf('.'));
+        } else if (javaVersion.indexOf('-') != -1) {
+            normalizedJavaVersion = javaVersion.substring(0, javaVersion.indexOf('-'));
         } else {
-            int dot = javaVersion.indexOf(".");
-            if (dot != -1) {
-                normalizedJavaVersion = javaVersion.substring(0, dot);
-            } else {
-                normalizedJavaVersion = javaVersion;
-            }
+            normalizedJavaVersion = javaVersion;
         }
         return Integer.parseInt(normalizedJavaVersion);
     }

--- a/awaitility/src/test/java/org/awaitility/core/JavaVersionDetectorTest.java
+++ b/awaitility/src/test/java/org/awaitility/core/JavaVersionDetectorTest.java
@@ -1,0 +1,75 @@
+package org.awaitility.core;
+
+import org.junit.Test;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.*;
+
+public class JavaVersionDetectorTest {
+
+    @Test
+    public void nullJavaVersionDefaultsTo8() {
+        assertThat(JavaVersionDetector.getJavaMajorVersion(null), equalTo(8));
+    }
+
+    @Test
+    public void emptyJavaVersionDefaultsTo8() {
+        assertThat(JavaVersionDetector.getJavaMajorVersion(""), equalTo(8));
+    }
+
+    @Test
+    public void javaVersionFor8() {
+        /*
+        docker run --rm -it eclipse-temurin:8-jdk bash
+        echo 'class Scratch {
+            public static void main(String[] args) {
+                System.out.println(System.getProperty("java.version"));
+            }
+        }' > Scratch.java
+        javac Scratch.java
+        java -cp . Scratch
+        */
+        assertThat(JavaVersionDetector.getJavaMajorVersion("1.8.0_402"), equalTo(8));
+    }
+
+    @Test
+    public void javaVersionFor11WithDot() {
+        /*
+        docker run --rm -it eclipse-temurin:11-jdk jshell
+        System.getProperty("java.version")
+         */
+        assertThat(JavaVersionDetector.getJavaMajorVersion("11.0.22"), equalTo(11));
+    }
+
+    @Test
+    public void javaVersionFor17WithDot() {
+        /*
+        docker run --rm -it eclipse-temurin:17-jdk jshell
+        System.getProperty("java.version")
+         */
+        assertThat(JavaVersionDetector.getJavaMajorVersion("17.0.10"), equalTo(17));
+    }
+
+    @Test
+    public void javaVersionFor21WithoutDot() {
+        /*
+        Until there is a first patch release, java.version reports a single number without dots.
+         */
+        assertThat(JavaVersionDetector.getJavaMajorVersion("21"), equalTo(21));
+    }
+
+    @Test
+    public void javaVersionFor21WithDot() {
+        /*
+        docker run --rm -it eclipse-temurin:21-jdk jshell
+        System.getProperty("java.version")
+         */
+        assertThat(JavaVersionDetector.getJavaMajorVersion("21.0.2"), equalTo(21));
+    }
+
+    @Test
+    public void javaVersionFor23ea() {
+        assertThat(JavaVersionDetector.getJavaMajorVersion("23-ea"), equalTo(23));
+    }
+
+}

--- a/awaitility/src/test/java/org/awaitility/core/JavaVersionDetectorTest.java
+++ b/awaitility/src/test/java/org/awaitility/core/JavaVersionDetectorTest.java
@@ -3,7 +3,7 @@ package org.awaitility.core;
 import org.junit.Test;
 
 import static org.hamcrest.MatcherAssert.assertThat;
-import static org.hamcrest.Matchers.*;
+import static org.hamcrest.Matchers.equalTo;
 
 public class JavaVersionDetectorTest {
 
@@ -69,7 +69,18 @@ public class JavaVersionDetectorTest {
 
     @Test
     public void javaVersionFor23ea() {
+        /*
+        OpenJDK EA reports java.version with ea suffix
+         */
         assertThat(JavaVersionDetector.getJavaMajorVersion("23-ea"), equalTo(23));
+    }
+
+    @Test
+    public void javaVersionFor23beta() {
+        /*
+        Temurin JDK EA reports java.version with beta suffix
+         */
+        assertThat(JavaVersionDetector.getJavaMajorVersion("23-beta"), equalTo(23));
     }
 
 }

